src/box/txn_limbo.c

```diff
@@ -773,38 +784,287 @@ txn_limbo_wait_empty(struct txn_limbo *limbo, double timeout)
 	return 0;
 }
 
-void
+/**
+ * Wait until all the limbo entries receive an lsn.
+ */
+static int
+txn_limbo_wait_persisted(struct txn_limbo *limbo)
+{
+	if (txn_limbo_is_empty(limbo))
+		return 0;
+
+	while (txn_limbo_last_entry(limbo)->lsn <= 0) {
+		fiber_cond_wait(&limbo->persist_cond);
+		if (fiber_is_cancelled()) {
+			diag_set(FiberIsCancelled);
+			return -1;
+		}
+	}
+	return 0;
+}
+
+/**
+ * Fill the reject reason with request data.
+ * The function is not reenterable, use with care.
+ */
+static char *
+reject_str(const struct synchro_request *req)
+{
+	const char *type_name = iproto_type_name(req->type);
+	static char prefix[128];
+
+	snprintf(prefix, sizeof(prefix), "RAFT: rejecting %s (%d) "
+		 "request from origin_id %u replica_id %u term %llu",
+		 type_name ? type_name : "UNKNOWN", req->type,
+		 req->origin_id, req->replica_id,
+		 (long long)req->term);
+
+	return prefix;
+}
+
+/**
+ * Common filter for any incoming packet.
+ */
+static int
+txn_limbo_filter_generic(struct txn_limbo *limbo,
+			 const struct synchro_request *req)
+{
+	assert(latch_is_locked(&limbo->promote_latch));
+
+	if (!limbo->validate)
+		return 0;
+
+	/*
+	 * Zero replica_id is allowed for PROMOTE packets only.
+	 */
+	if (req->replica_id == REPLICA_ID_NIL) {
+		if (req->type != IPROTO_RAFT_PROMOTE) {
+			say_error("%s. Zero replica_id detected",
+				  reject_str(req));
+			diag_set(ClientError, ER_UNSUPPORTED, "Replication",
+				 "synchronous requests with zero replica_id");
+			return -1;
+		}
+	}
+	if (req->replica_id != limbo->owner_id) {
+		/*
+		 * Incoming packets should esteem limbo owner,
+		 * if it doesn't match it means the sender
+		 * missed limbo owner migrations and is out of date.
+		 */
+		say_error("%s. Limbo owner mismatch, owner_id %u",
+			  reject_str(req), limbo->owner_id);
+		diag_set(ClientError, ER_SPLIT_BRAIN,
+			 "got a request from a foreign synchro queue owner");
+		return -1;
+	}
+
+	return 0;
+}
+
+/**
+ * A common filter for all synchro requests, checking that request operates
+ * over a valid lsn range.
+ */
+static int
+filter_queue_boundaries(struct txn_limbo *limbo,
+			const struct synchro_request *req)
+{
+	int64_t lsn = req->lsn;
+	/*
+	 * Easy case - processed LSN matches the new one which comes inside
+	 * request, everything is consistent.
+	 */
+	if (limbo->confirmed_lsn == lsn)
+		return 0;
+
+	/*
+	 * Explicit split brain situation. Request comes in with an old LSN
+	 * which we've already processed.
+	 */
+	if (limbo->confirmed_lsn > lsn) {
```

Comment:
Not all `if`s correspond to a reproducible scenario. Some of them (like the ones with lsn == 0, zero term, zero replica id, CONFIRM for lsn == confirmed_lsn) could only be reproduced with fuzzing.
Since these situations are not really a split brain, but some other issue, I tried to raise `ER_UNSUPPORTED` and not `ER_SPLIT_BRAIN` for them.

But speaking of all the `ER_SPLIT_BRAIN` errors, they are covered. At least they were until the todays push. Let's see how it goes now. I checked via coveralls.